File: csharp/src/SlimShader/Chunks/Shex/Tokens/ResourceReturnTypeToken.cs
using SlimShader.Chunks.Common;
using SlimShader.Util;

namespace SlimShader.Chunks.Shex.Tokens
{
	/// <summary>
	/// Resource Return Type Token (ResourceReturnTypeToken) (used in resource
	/// declaration statements)
	///
	/// [03:00] D3D10_SB_RESOURCE_RETURN_TYPE for component X
	/// [07:04] D3D10_SB_RESOURCE_RETURN_TYPE for component Y
	/// [11:08] D3D10_SB_RESOURCE_RETURN_TYPE for component Z
	/// [15:12] D3D10_SB_RESOURCE_RETURN_TYPE for component W
	/// [31:16] Reserved, 0
	/// </summary>
	public class ResourceReturnTypeToken
	{
		public ResourceReturnType X { get; internal set; }
		public ResourceReturnType Y { get; internal set; }
		public ResourceReturnType Z { get; internal set; }
		public ResourceReturnType W { get; internal set; }

		public static ResourceReturnTypeToken Parse(BytecodeReader reader)
		{
			var token = reader.ReadUInt32();
			return new ResourceReturnTypeToken
			{
				X = token.DecodeValue<ResourceReturnType>(00, 03),
				Y = token.DecodeValue<ResourceReturnType>(04, 07),
				Z = token.DecodeValue<ResourceReturnType>(08, 11),
				W = token.DecodeValue<ResourceReturnType>(12, 15)
			};
		}

		public override string ToString()
		{
			return string.Format("{0},{1},{2},{3}",
				X.GetDescription(),
				Y.GetDescription(),
				Z.GetDescription(),
				W.GetDescription());
		}
	}
}

File: src/cpp/Source/Chunks/Shex/Tokens/ResourceReturnTypeToken.cpp
#include "PCH.h"
#include "ResourceReturnTypeToken.h"

#include "Decoder.h"

using namespace std;
using namespace SlimShader;

ResourceReturnTypeToken ResourceReturnTypeToken::Parse(BytecodeReader& reader)
{
	auto token = reader.ReadUInt32();
	ResourceReturnTypeToken result;
	result._x = DecodeValue<ResourceReturnType>(token, 0, 3);
	result._y = DecodeValue<ResourceReturnType>(token, 4, 7);
	result._z = DecodeValue<ResourceReturnType>(token, 8, 11);
	result._w = DecodeValue<ResourceReturnType>(token, 12, 15);
	return result;
}

ResourceReturnType ResourceReturnTypeToken::GetX() const { return _x; }
ResourceReturnType ResourceReturnTypeToken::GetY() const { return _y; }
ResourceReturnType ResourceReturnTypeToken::GetZ() const { return _z; }
ResourceReturnType ResourceReturnTypeToken::GetW() const { return _w; }

ostream& SlimShader::operator<<(ostream& out, const ResourceReturnTypeToken& value)
{
	out << boost::format("%s,%s,%s,%s")
		% ToString(value._x)
		% ToString(value._y)
		% ToString(value._z)
		% ToString(value._w);
	return out;
}